File: velocity-position-rotation_mode/state_machine.py


class State_Machine:

    def __init__(self, states = {}):
        self.state_dict = states
        self.current_state = None

    def add_state(self, name, state_class):
        self.state_dict[name] = state_class

    def set_state(self, name):
        if self.current_state != name:
            print(f"transitioning to {name}")

        self.current_state = name


    def get_state(self):
        return self.current_state

    def run(self):
        state_class  = self.state_dict.get(self.current_state)
        if not state_class:
            raise Exception(f"state {self.current_state} not found", self.state_dict)
        self.set_state(state_class.run())

File: kotlin/src/main/java/frc/robot/StateMachine.kt

package frc.robot

class StateMachine(states : Map<String, State>, smName : String = "") {

  private val stateDict = states
  private var currentState : String? = null
  private val m_smName = smName

  public fun setState(name : String) {
    if (currentState != name) {
      println("$m_smName: transitioning to $name")
    }
    currentState = name
  }

  public fun getState() : String? {
      return currentState
  }

  public fun run() {
    if (!stateDict.containsKey(currentState)) {
      throw Exception("state $currentState not found")
    }

    val stateClass = stateDict.get(currentState)

    if (stateClass != null) {
      setState(stateClass.run())
    }
    
  }

}